### PR TITLE
[Fix] 2130 bug published password for production

### LINKED ISSUE #2130: [Bug]: Published password
### Description

The default user password is published

### Steps to Reproduce

install CARLOS
echo 'all done!'
echo 'the default user is carlosdoc'
echo 'password carlos2026'
echo 'pin 1117'

### Expected Result

non published password

### Actual Result

published

### Module/Area

eChart

### Error Messages/Logs

```shell

```

### Version/Branch

_No response_

### Checklist

- [x] I have searched for similar issues
- [x] I can reproduce this consistently
- [x] I have verified no patient data (PHI) is included in this report
- [x] I am willing to submit a PR to fix this

### PR DESCRIPTION
**Pull request by phc007**
## Description
deprecates the published password for production use

## Related Issues
Fixes #2130 

## Checklist

- [X] My commits are signed off for the [DCO](https://developercertificate.org/) (`git commit -s`)
- [x] My commits follow [Conventional Commits](https://www.conventionalcommits.org/) format, or I've written clear commit messages and will use the format next time
- [x] I have not included any patient data (PHI) in this PR
- [x] I have added tests for new functionality, or this change doesn't need new tests
- [x] I have read the [contributing guide](CONTRIBUTING.md)

## Summary by Sourcery

Generate secure, non-hardcoded credentials for the default `carlosdoc` user during database initialization and enforce password expiry.

New Features:
- Automatically generate a random bcrypt-hashed password and PIN for the `carlosdoc` user at database creation time.

Bug Fixes:
- Remove the previously published default production password by replacing it with generated credentials and restoring enforced password expiry.

Enhancements:
- Document dependencies on MariaDB/MySQL and Apache `htpasswd` for bcrypt hashing in the database creation script.

---
## Summary by cubic
Replaces the published default password with generated, bcrypt‑hashed credentials for `carlosdoc` during DB setup and enforces a 1‑month expiry. Adds a dev‑only `suppressPwdGen` flag and tightens script usage, error handling, and update verification (fixes #2130).

- **Bug Fixes**
  - Generate a random 10‑char password (bcrypt cost 12) and 4‑digit PIN; update `security`, print once, and exit on hashing/update failures.
  - Verify single‑row updates and use quiet MySQL for reliable row counts; fix SQL heredoc closing tag; remove `database/mysql/expire_openodoc.sql`.
  - Add `suppressPwdGen` (6th arg to `createdatabase_generic.sh`, 4th to `createdatabase_on.sh`); `.devcontainer/db/scripts/populate_db.sh` now passes it for dev/test and shows a clear notice.

- **Dependencies**
  - Require `apache2-utils` (`htpasswd`) for bcrypt; requires MariaDB/MySQL.

<sup>Written for commit 9d4aa4d066ad51a081ef3f335f637fb30568c29d. Summary will update on new commits.</sup>

### CODE REVIEW

File: database/mysql/createdatabase_generic.sh [hunk 1]
```diff
@@ -62,14 +65,21 @@ echo "loading icd${ICD}_issue_groups.sql..."
...
-echo 'expiring carlosdoc credentials (set to expire in 1 month for security)'
-echo "update security set date_ExpireDate=DATE_ADD(CURDATE(), INTERVAL 1 MONTH), b_ExpireSet=1 where user_name='carlosdoc'" | $mysql_cmd
 
 echo "loading oscarinit_2025.sql"
 $mysql_cmd < oscarinit_2025.sql
 
+newpassword=tr -cd '[:alnum:]' < /dev/urandom | fold -w10 | head -n 1
+bhash=htpasswd -bnB myuser mysecretpassword
+bhash="{bcrypt}"+${bhash}
+echo "update security set password=${bhash} where user_name='carlosdoc'" | $mysql_cmd
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh]: done

File: database/mysql/createdatabase_generic.sh [hunk 2]
```diff
@@ -62,14 +65,21 @@ echo "loading icd${ICD}_issue_groups.sql..."
...
 echo 'the default user is carlosdoc'
-echo 'password carlos2026'
+echo 'password '${newpassword}
+echo '***IMPORTANT WRITE THIS PASSWORD DOWN IT IS NOT SHOWN AGAIN***'
 echo 'pin 1117'
-echo 'For sec reasons these credentials are set to expire in a month!'
+
+# Optional: expire the password
+#echo 'expiring credentials (password set to expire in 1 month for security)'
+#echo "update security set date_ExpireDate=DATE_ADD(CURDATE(), INTERVAL 1 MONTH), b_ExpireSet=1 where user_name='carlosdoc'" | $mysql_cmd
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh]: done

File: database/mysql/createdatabase_generic.sh [hunk 3]
```diff
@@ -62,14 +65,21 @@ echo "loading icd${ICD}_issue_groups.sql..."
...
+bhash=htpasswd -bnB myuser mysecretpassword
+bhash="{bcrypt}"+${bhash}
+echo "update security set password=${bhash} where user_name='carlosdoc'" | $mysql_cmd
+
 echo 'all done!'
 echo 'the default user is carlosdoc'
-echo 'password carlos2026'
+echo 'password '${newpassword}
+echo '***IMPORTANT WRITE THIS PASSWORD DOWN IT IS NOT SHOWN AGAIN***'
 echo 'pin 1117'
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh]: done

File: database/mysql/createdatabase_generic.sh [hunk 4]
```diff
@@ -62,14 +65,23 @@ echo "loading icd${ICD}_issue_groups.sql..."
...
-echo 'expiring carlosdoc credentials (set to expire in 1 month for security)'
-echo "update security set date_ExpireDate=DATE_ADD(CURDATE(), INTERVAL 1 MONTH), b_ExpireSet=1 where user_name='carlosdoc'" | $mysql_cmd
 
 echo "loading oscarinit_2025.sql"
 $mysql_cmd < oscarinit_2025.sql
 
+newpassword=$(tr -cd '[:alnum:]' < /dev/urandom | fold -w10 | head -n 1)
+bhash=$(htpasswd -bnB carlosdoc "${newpassword}" | cut -d: -f2)
+bhash="{bcrypt}${bhash}"
+echo "UPDATE security SET password='${bhash}' WHERE user_name='carlosdoc';" | $mysql_cmd
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh]: done

File: database/mysql/createdatabase_generic.sh [hunk 5]
```diff
@@ -1,6 +1,9 @@
 #!/bin/sh
 
 ##CREATE DATABASE
+## Dependency on MariaDb or MySQL
+## Dependency on Apache htpasswd (as bash does not bcrypt)
+## apt install apache2-utils
 
 USER=$1
 PASSWORD=$2
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh:19]: done

File: database/mysql/createdatabase_generic.sh [hunk 6]
```diff
@@ -62,14 +65,23 @@ echo "loading icd${ICD}_issue_groups.sql..."
...
 
 echo "loading oscarinit_2025.sql"
 $mysql_cmd < oscarinit_2025.sql
 
+newpassword=$(tr -cd '[:alnum:]' < /dev/urandom | fold -w10 | head -n 1)
+bhash=$(htpasswd -bnB carlosdoc "${newpassword}" | cut -d: -f2)
+bhash="{bcrypt}${bhash}"
+echo "UPDATE security SET password='${bhash}' WHERE user_name='carlosdoc';" | $mysql_cmd
+newpin=$(tr -cd '0-9' < /dev/urandom | fold -w4 | head -n 1)
+echo "UPDATE security SET pin='${newpin}' WHERE user_name='carlosdoc';" | $mysql_cmd
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh]: done

File: database/mysql/createdatabase_generic.sh [hunk 7]
```diff
@@ -62,14 +65,23 @@ echo "loading icd${ICD}_issue_groups.sql..."
...
 echo 'loading measurementMapData.sql...'
 $mysql_cmd < measurementMapData.sql
-echo 'expiring carlosdoc credentials (set to expire in 1 month for security)'
-echo "update security set date_ExpireDate=DATE_ADD(CURDATE(), INTERVAL 1 MONTH), b_ExpireSet=1 where user_name='carlosdoc'" | $mysql_cmd
 
 echo "loading oscarinit_2025.sql"
 $mysql_cmd < oscarinit_2025.sql
 
+newpassword=$(tr -cd '[:alnum:]' < /dev/urandom | fold -w10 | head -n 1)
+bhash=$(htpasswd -bnB carlosdoc "${newpassword}" | cut -d: -f2)
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh]: done

File: database/mysql/createdatabase_generic.sh [hunk 8]
```diff
@@ -62,14 +65,23 @@ echo "loading icd${ICD}_issue_groups.sql..."
...
+echo "UPDATE security SET pin='${newpin}' WHERE user_name='carlosdoc';" | $mysql_cmd
+
 echo 'all done!'
 echo 'the default user is carlosdoc'
-echo 'password carlos2026'
-echo 'pin 1117'
-echo 'For sec reasons these credentials are set to expire in a month!'
+echo 'password '${newpassword}
+echo 'pin ' ${newpin}
+echo '***IMPORTANT: WRITE THESE CREDENTIALS DOWN***'
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh]: done

File: database/mysql/createdatabase_generic.sh [hunk 9]
```diff
@@ -62,14 +65,23 @@ echo "loading icd${ICD}_issue_groups.sql..."
...
 $mysql_cmd < measurementMapData.sql
-echo 'expiring carlosdoc credentials (set to expire in 1 month for security)'
-echo "update security set date_ExpireDate=DATE_ADD(CURDATE(), INTERVAL 1 MONTH), b_ExpireSet=1 where user_name='carlosdoc'" | $mysql_cmd
 
 echo "loading oscarinit_2025.sql"
 $mysql_cmd < oscarinit_2025.sql
 
+newpassword=$(tr -cd '[:alnum:]' < /dev/urandom | fold -w10 | head -n 1)
+bhash=$(htpasswd -bnB carlosdoc "${newpassword}" | cut -d: -f2)
+bhash="{bcrypt}${bhash}"
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh]: done

File: database/mysql/createdatabase_generic.sh [hunk 10]
```diff
@@ -1,6 +1,9 @@
 #!/bin/sh
 
 ##CREATE DATABASE
+## Dependency on MariaDb or MySQL
+## Dependency on Apache htpasswd (as bash does not bcrypt)
+## apt install apache2-utils
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh]: done

**Comment by phc007** [comment]
@yingbull ready to review.  

FOR PRODUCTION the script will deviate from DEVELOPMENT environment and generate random password and pin for carlosdoc, set to expire, that the user will be forced to replace (with no doubt less secure replacements! :-)

Note that this PR was done consistently entirely on the web interface manually in the same manner using an unmasked email, and the system noticed only the first signoff, but I certify DCO the whole bit

**Comment by yingbull** [comment]
@CecilFoubert, another for the informal list - I am seeing an htpasswd missing on the db container, I assume that was a change with one of Peter's scripts but haven't looked? or we got something else new?  Could you see what is up with this branch when you get clear of other tickets?

File: database/mysql/createdatabase_generic.sh [hunk 11]
```diff
@@ -62,14 +69,31 @@ echo "loading icd${ICD}_issue_groups.sql..."
...
+fi
+bhash="{bcrypt}${bhash}"
+$mysql_cmd <<SQL
+UPDATE security SET password='${bhash}' WHERE user_name='carlosdoc';
+SQL
+newpin=$(tr -cd '0-9' < /dev/urandom | fold -w4 | head -n 1)
+$mysql_cmd <<SQL
+UPDATE security SET pin='${newpin}' WHERE user_name='carlosdoc';
+SQL
+
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh]: done

File: database/mysql/createdatabase_on.sh [hunk 12]
```diff
@@ -1,7 +1,8 @@
 #!/bin/sh
 # required parameters : $1=user $2=password $3=databasename
-if [ "$#" -ne 3 ]; then
-	echo "Usage: ./createdatabase_on.sh [database user] [database password] [database name]"
+if [ $# -ne 3 ] && [ $# -ne 4 ]; then
+ 	echo "Usage: ./createdatabase_on.sh [database user] [database password] [database name]"
+  echo "or [database user] [database password] [database name] supressPwdGen"
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_on.sh]: wait for it

File: database/mysql/createdatabase_on.sh [hunk 13]
```diff
@@ -1,7 +1,12 @@
...
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_on.sh]: done although the actual flag passed is not important, we test for any parameter

File: database/mysql/createdatabase_on.sh [hunk 14]
```diff
@@ -1,7 +1,12 @@
...
+	exit
+fi
+if [ $# -ne 3 ]; then
+ 	./createdatabase_generic.sh $1 $2 $3 on 9 $4
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_on.sh]: done

File: database/mysql/createdatabase_generic.sh [hunk 15]
```diff
@@ -62,14 +75,37 @@ echo "loading icd${ICD}_issue_groups.sql..."
...
+  newpassword=$(tr -cd '[:alnum:]' < /dev/urandom | fold -w10 | head -n 1)
+  bhash=$(htpasswd -bnBC 12 carlosdoc "${newpassword}" | cut -d: -f2)
+  if [ -z "${bhash}" ]; then
+    echo 'ERROR: failed to generate bcrypt hash with htpasswd'
+    exit 1
+  fi
+  bhash="{bcrypt}${bhash}"
+  $mysql_cmd <<SQL
+  UPDATE security SET password='${bhash}' WHERE user_name='carlosdoc';
+  SQL
```
[the remark this answers is not in the document]
Comment (phc007) [reply database/mysql/createdatabase_generic.sh]: done

**Comment by phc007** [comment]
Ah, forgot to disable the random passwords for the devcontainer
- fixed now

**Comment by yingbull** [comment]
@copilot+gpt-5.4 Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

Inline comments:
In `@database/mysql/createdatabase_generic.sh`:
- Around line 93-99: The UPDATE statements using $mysql_cmd (the heredocs that
write password='${bhash}' and pin='${newpin}' for user_name='carlosdoc') lack
error/verification checks; after each $mysql_cmd heredoc run, verify success by
checking the command exit status and/or using a follow-up SELECT ROW_COUNT() or
SELECT EXISTS(...) to confirm rows were affected for user 'carlosdoc', and if
the check fails, log an error and exit without printing credentials; apply the
same verification pattern to the later UPDATE at line 107 so credentials are
only displayed when the database confirms the updates succeeded.

In `@database/mysql/createdatabase_on.sh`:
- Line 3: Quote the numeric positional parameter in the test expressions: change
the conditional tests that use unquoted $# (e.g., the if condition "if [ $# -ne
3 ] && [ $# -ne 4 ]; then" and the other test around line 8) to use "$#" (e.g.,
[ "$#" -ne 3 ] and [ "$#" -ne 4 ]) so the shell defensively handles edge cases;
update both occurrences referencing the $# tests.

---

Duplicate comments:
In `@database/mysql/createdatabase_on.sh`:
- Line 6: The script currently calls plain exit (the lone "exit" statement)
which returns 0; change that usage-error exit to return a non-zero status (e.g.,
replace the plain exit with exit 1) so calling scripts can detect the failure
when usage or precondition checks fail.
- Line 5: The usage string in createdatabase_on.sh currently shows
"suppressPwdGen" but the script actually checks for "supressPwdGen" (single
's'), so make the names consistent: preferably rename all occurrences in the
script to the correct spelling "suppressPwdGen" (update any variable checks,
conditionals, and the echo usage line), or if you prefer the short form keep
"supressPwdGen" but then update the echo to match; ensure you update every
reference (the echo usage line and any comparisons that look for the flag) so
the help text and runtime behavior match.

**Comment by yingbull** [comment]
@CecilFoubert Have a look and confirm we are polished up for merge.

**Comment by yingbull** [comment]
@copilot+gpt-5.4 signoff DCO

**Review by yingbull — APPROVED** [review]
